### PR TITLE
Improve CI and packaging

### PR DESCRIPTION
**Pull request by JeanChristopheMorinPerso**
This pull request improves the general packaging of pyavb and also improves the CI.

Packaging improvements:
* Wheels produced when `PYAVB_BUILD_EXT` is set to 0 are now marked as pure and universal and compatible with both Python 2 and 3. This allows to have a single wheel with the idea that it's unnecessary to have python and platform specific wheels when the content is just pure python code. See the `bdist_wheel` section in the `setup.cfg` file to see how I did this.
* [python_requires](https://packaging.python.org/en/latest/guides/distributing-packages-using-setuptools/#python-requires) metadata is now set to let `pip` (and any other installers) which Python version the library supports. This is extremely important when a library starts to drop support for Python 2 and any other versions of Python.
* Wheels are now built using [cibuildwheel](https://cibuildwheel.readthedocs.io/en/stable/). This allows to have known working wheels. This is because it takes care of compiling in vetted environments:
    * on Linux it uses docker images while on mac and Windows it builds on the host (because of their lack of docker support).
    * Wheels are "repaired", which means it creates wheels, then inspects them to make sure that it is self-contained and doesn't rely on any system level libraries that could cause problems between RedHat and Ubuntu for example.
* Build dependencies are now provided via a [pyproject.toml](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/) file. This means that if you `pip install .`, it will create a virtualenv with the dependencies set in pyproject.toml to build the package. Again, this is important for reproducibility.

CI improvements:
* Wheels (for both compiled and non-compiled avb) are first created, then uploaded as artifacts, then the tests jobs download the artifacts and install the wheels. The tests are run against the installed wheels. This allows to verify that the wheels are working as expected (for example, if some files are missing, the tests would probably fail). It gives extra confidence that wheels can be released.
* I changed the layout of the code to use what's called a src layout. That is the code lives under the `src` directory. This avoids CI to accidentally tests the source code instead of the installed code. You can see some explanations at https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#src-layout and https://hynek.me/articles/testing-packaging/.
* Pure and non-pure (compiled) wheels are tested. This ensures that pure (non-compiled) wheels continue to work (since the default is to use Cython if it's available).
* Python 2.7 on Windows: The MSVC compiler is no more downloaded from the Internet Archive. Wheels are now compiled using what's available in the GitHub Actions windows runners. To achieve this, I had to set the [/MT](https://docs.microsoft.com/en-us/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-170) compiler flag. This method has been extensively battle tested by OTIO.

I manually tested the Python 2.7 wheel on Windows using a clean VM. Testing on a clean VM was required in this case because I wanted to be 100% that the latest runtime libraries were not present to ensure that the `/MT` flag did its job. I also confirmed the that `pyd` only linked to Python 2.7 DLL and nothing else, using `dumpbin /dependents C:\path\to'_ext.pyd` ([dumpbin](https://docs.microsoft.com/en-us/cpp/build/reference/dumpbin-reference?view=msvc-170) can be somewhat seen as the equivalent of `ldd` but for Windows).

But I still encourage you to test the wheels, just in case I missed something.

I hope this little wall of text isn't too much to digest... If there is anything, please let me know and I'll be happy to walk you through the changes or answer any questions or concerns.

### CODE REVIEW

**Comment by JeanChristopheMorinPerso** [comment]
BTW, you will have to approve the workflow so that it can run.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[build-system]
+requires = [
+    "setuptools>=42", "cython==0.29.30"
```
Comment (JeanChristopheMorinPerso) [inline pyproject.toml:3]: Notice how the Cython version is baked here.

File: setup.cfg [hunk 2]
```diff
@@ -1,3 +1,8 @@
 [metadata]
 long_description = file: README.rst
 long_description_content_type = text/x-rst
+
+[bdist_wheel]
+# Let a wheel be compatible with both Python 2 and 3
+# when Cython is not used.
+universal=1
```
Comment (JeanChristopheMorinPerso) [inline setup.cfg:8]: This is how universal wheels can be created.

File: setup.py [hunk 3]
```diff
@@ -33,6 +35,28 @@
...
+    def build_extensions(self):
+        for ext in self.extensions:
+            flags = []
+            if self.compiler.compiler_type == 'msvc':
+                if sys.version_info[0] == 2:
+                    # If we are compiling for Python 2.7 on Windows, statically link
+                    # the runtime libraries. THis will allow to compile using an
+                    # msvc greater than the one used to compile Python 2.7.
+                    # See See https://docs.microsoft.com/en-us/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-160.
+                    flags.append('/MT')
```
Comment (JeanChristopheMorinPerso) [inline setup.py:50]: The `/MT` flag is here, just for Python 2 and the MSVC compiler.

File: Makefile [hunk 4]
```diff
@@ -12,11 +12,13 @@ python-version:
...
-	- find avb -name '*.dylib' -delete
-	- find avb -name '*.pyd' -delete
-	- find avb -name '*.dll' -delete
-	- find avb -name '*.pyc' -delete
+	- find src/avb -name '*.so' -delete
+	- find src/avb -name '*.dylib' -delete
+	- find src/avb -name '*.pyd' -delete
+	- find src/avb -name '*.dll' -delete
+	- find src/avb -name '*.pyc' -delete
+	- rm -f src/avb/_ext.cpp
```
Comment (JeanChristopheMorinPerso) [inline Makefile:20]: I added the removal of this file as it is auto-generated by Cython on every build.

**Comment by markreidvfx** [comment]
Thanks for being so detailed! I see the latest-release step failed, but that shouldn't be a issue as far a I can tell? I'll test it out locally and merge it if I don't find any issues. `publish.yml` will also need to be updated with the new artifacts names but I can do that.

**Comment by JeanChristopheMorinPerso** [comment]
Yeah, I was expecting the `latest-release` to fail. It fails because PRs coming from forks don't have write access to the target repo (for quite obvious reasons... You don't want random people to create PRs that would damage the repo).

As for the release pipeline, I completely forgot about it. I'll push a commit to use the new artifact name.

**Comment by markreidvfx** [comment]
do the `pure-wheels` need to uploaded to pypi too?

**Comment by JeanChristopheMorinPerso** [comment]
In theory no. I'd have to do some testing and reading to see if it's common to have both pure and non pure wheels in PyPI. So I went with just uploading the Cython wheels for now.

**Comment by JeanChristopheMorinPerso** [comment]
So if you want, give me some time and I'll come back to tell if we can upload both the pure and non pure wheels.

**Comment by markreidvfx** [comment]
I think it will be fine for now as long as we upload the `sdist`.

**Comment by JeanChristopheMorinPerso** [comment]
After looking at what Cython do themselves, they publish compiled and non compiled wheels + sdist:

```
$ curl -s https://pypi.org/pypi/cython/0.29.30/json | jq '.releases["0.29.30"][].filename' -r
Cython-0.29.30-cp27-cp27m-manylinux_2_5_i686.manylinux1_i686.whl
Cython-0.29.30-cp27-cp27m-manylinux_2_5_x86_64.manylinux1_x86_64.whl
Cython-0.29.30-cp27-cp27mu-manylinux_2_5_i686.manylinux1_i686.whl
Cython-0.29.30-cp27-cp27mu-manylinux_2_5_x86_64.manylinux1_x86_64.whl
Cython-0.29.30-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.manylinux_2_24_aarch64.whl
Cython-0.29.30-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl
Cython-0.29.30-cp310-cp310-manylinux_2_5_i686.manylinux1_i686.manylinux_2_24_i686.whl
Cython-0.29.30-cp310-cp310-musllinux_1_1_x86_64.whl
Cython-0.29.30-cp311-cp311-manylinux_2_17_aarch64.manylinux2014_aarch64.manylinux_2_24_aarch64.whl
Cython-0.29.30-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl
Cython-0.29.30-cp311-cp311-manylinux_2_5_i686.manylinux1_i686.manylinux_2_24_i686.whl
Cython-0.29.30-cp311-cp311-musllinux_1_1_x86_64.whl
Cython-0.29.30-cp35-cp35m-manylinux_2_5_i686.manylinux1_i686.whl
Cython-0.29.30-cp35-cp35m-manylinux_2_5_x86_64.manylinux1_x86_64.whl
Cython-0.29.30-cp36-cp36m-manylinux_2_17_aarch64.manylinux2014_aarch64.manylinux_2_24_aarch64.whl
Cython-0.29.30-cp36-cp36m-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl
Cython-0.29.30-cp36-cp36m-manylinux_2_5_i686.manylinux1_i686.manylinux_2_24_i686.whl
Cython-0.29.30-cp36-cp36m-manylinux_2_5_i686.manylinux1_i686.whl
Cython-0.29.30-cp36-cp36m-manylinux_2_5_x86_64.manylinux1_x86_64.whl
Cython-0.29.30-cp36-cp36m-musllinux_1_1_x86_64.whl
Cython-0.29.30-cp37-cp37m-manylinux_2_17_aarch64.manylinux2014_aarch64.manylinux_2_24_aarch64.whl
Cython-0.29.30-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl
Cython-0.29.30-cp37-cp37m-manylinux_2_5_i686.manylinux1_i686.manylinux_2_24_i686.whl
Cython-0.29.30-cp37-cp37m-manylinux_2_5_i686.manylinux1_i686.whl
Cython-0.29.30-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.whl
Cython-0.29.30-cp37-cp37m-musllinux_1_1_x86_64.whl
Cython-0.29.30-cp38-cp38-manylinux_2_17_aarch64.manylinux2014_aarch64.manylinux_2_24_aarch64.whl
Cython-0.29.30-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl
Cython-0.29.30-cp38-cp38-manylinux_2_5_i686.manylinux1_i686.manylinux_2_24_i686.whl
Cython-0.29.30-cp38-cp38-manylinux_2_5_i686.manylinux1_i686.whl
Cython-0.29.30-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.whl
Cython-0.29.30-cp38-cp38-musllinux_1_1_x86_64.whl
Cython-0.29.30-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.manylinux_2_24_aarch64.whl
Cython-0.29.30-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl
Cython-0.29.30-cp39-cp39-manylinux_2_5_i686.manylinux1_i686.manylinux_2_24_i686.whl
Cython-0.29.30-cp39-cp39-manylinux_2_5_i686.manylinux1_i686.whl
Cython-0.29.30-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.whl
Cython-0.29.30-cp39-cp39-musllinux_1_1_x86_64.whl
Cython-0.29.30-py2.py3-none-any.whl
Cython-0.29.30.tar.gz
```

So I pushed another commit to do that too. Thanks for asking the question!

**Pull request merged**